rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -356,6 +356,101 @@ pub struct SystemState {
+}
+
+/// A wrapper around the different canister statuses.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Default)]
+pub enum OnLowWasmMemoryHookStatus {
+    #[default]
+    ConditionNotSatisfied,
+    Ready,
+    Executed,
+}
+
+impl From<&OnLowWasmMemoryHookStatus> for pb::OnLowWasmMemoryHookStatus {
+    fn from(item: &OnLowWasmMemoryHookStatus) -> Self {
+        match *item {
+            OnLowWasmMemoryHookStatus::ConditionNotSatisfied => {
+                pb::OnLowWasmMemoryHookStatus::ConditionNotSatisfied
+            }
+            OnLowWasmMemoryHookStatus::Ready => pb::OnLowWasmMemoryHookStatus::Ready,
+            OnLowWasmMemoryHookStatus::Executed => pb::OnLowWasmMemoryHookStatus::Executed,
+        }
+    }
+}
+
+impl TryFrom<pb::OnLowWasmMemoryHookStatus> for OnLowWasmMemoryHookStatus {
+    type Error = ProxyDecodeError;
+
+    fn try_from(value: pb::OnLowWasmMemoryHookStatus) -> Result<Self, Self::Error> {
+        match value {
+            pb::OnLowWasmMemoryHookStatus::Unspecified => Err(ProxyDecodeError::ValueOutOfRange {
+                typ: "OnLowWasmMemoryHookStatus",
+                err: format!(
+                    "Unexpected value of status of on low wasm memory hook: {:?}",
+                    value
+                ),
+            }),
+            pb::OnLowWasmMemoryHookStatus::ConditionNotSatisfied => {
+                Ok(OnLowWasmMemoryHookStatus::ConditionNotSatisfied)
+            }
+            pb::OnLowWasmMemoryHookStatus::Ready => Ok(OnLowWasmMemoryHookStatus::Ready),
+            pb::OnLowWasmMemoryHookStatus::Executed => Ok(OnLowWasmMemoryHookStatus::Executed),
+        }
+    }
+}
+
+impl OnLowWasmMemoryHookStatus {
+    pub fn update(
+        &mut self,
+        wasm_memory_threshold: NumBytes,
+        memory_allocation: Option<NumBytes>,
+        used_stable_memory: NumBytes,
+        used_wasm_memory: NumBytes,
+    ) {
+        // The maximum Wasm memory occupied by the canister is 4 GiB.
+        let max_wasm_capacity = NumBytes::new(4 * 1024 * 1024 * 1024);
+
+        // If the canister has memory allocation, then it maximum allowed Wasm memory
+        // can be calculated as min(memory_allocation - used_stable_memory, 4 GiB).
+        let wasm_capacity = if let Some(memory_allocation) = memory_allocation {
+            debug_assert!(
+                used_stable_memory <= memory_allocation,
+                "Used stable memory: {:?}, is larger than memory allocation: {:?}.",
+                used_stable_memory,
+                memory_allocation
+            );
+            std::cmp::min(memory_allocation - used_stable_memory, max_wasm_capacity)
+        } else {
+            max_wasm_capacity
+        };
+
+        debug_assert!(
+            used_wasm_memory <= wasm_capacity,
+            "Used wasm memory: {:?}, is larger than maximal possible: {:?}.",
+            used_wasm_memory,
+            wasm_capacity
+        );
```

Comment:
`//rs/execution_environment:execution_environment_test` is failing here:
`Used wasm memory: 4294967716, is larger than maximal possible: 4294967296.`